File: apps/syn-cli/src/syn_cli/commands/observe.py
"""Observability commands — tool timelines and token breakdowns."""

from __future__ import annotations

import typer
from rich.panel import Panel
from rich.table import Table

from syn_cli._output import console, format_cost, format_duration, format_tokens
from syn_cli.commands._api_helpers import api_get

app = typer.Typer(
    name="observe",
    help="Observability data for sessions — tool timelines and token metrics",
    no_args_is_help=True,
)


@app.command("tools")
def tool_timeline(
    session_id: str = typer.Argument(..., help="Session ID"),
    limit: int = typer.Option(100, "--limit", "-n", help="Max results", min=1, max=500),
) -> None:
    """Show tool execution timeline for a session."""
    data = api_get(
        f"/observability/sessions/{session_id}/tools",
        params={"limit": limit},
    )

    executions = data.get("executions", [])
    if not executions:
        console.print("[dim]No tool executions found for this session.[/dim]")
        return

    console.print(
        f"[dim]Session: {data.get('session_id', session_id)}  |  Total: {data.get('total_executions', len(executions))}[/dim]"
    )

    table = Table(title="Tool Timeline")
    table.add_column("Tool", style="cyan")
    table.add_column("Type")
    table.add_column("Duration", justify="right")
    table.add_column("Status")

    for ex in executions:
        dur = format_duration(ex.get("duration_ms", 0)) if ex.get("duration_ms") else "-"
        ok = "[green]ok[/green]" if ex.get("success", True) else "[red]fail[/red]"
        table.add_row(
            ex.get("tool_name", "-"),
            ex.get("operation_type", "-"),
            dur,
            ok,
        )
    console.print(table)


@app.command("tokens")
def token_metrics(
    session_id: str = typer.Argument(..., help="Session ID"),
) -> None:
    """Show token breakdown for a session."""
    data = api_get(f"/observability/sessions/{session_id}/tokens")

    panel_text = (
        f"[bold]Session:[/bold] {data.get('session_id', session_id)}\n"
        f"[bold]Input Tokens:[/bold] {format_tokens(data.get('input_tokens', 0))}\n"
        f"[bold]Output Tokens:[/bold] {format_tokens(data.get('output_tokens', 0))}\n"
        f"[bold]Total Tokens:[/bold] {format_tokens(data.get('total_tokens', 0))}\n"
        f"[bold]Cache Creation:[/bold] {format_tokens(data.get('cache_creation_tokens', 0))}\n"
        f"[bold]Cache Read:[/bold] {format_tokens(data.get('cache_read_tokens', 0))}\n"
        f"[bold]Cost:[/bold] {format_cost(data.get('total_cost_usd', '0'))}"
    )
    console.print(Panel(panel_text, title="[cyan]Token Metrics[/cyan]", border_style="cyan"))

File: apps/syn-cli-node/src/commands/observe.ts
/**
 * Observability commands — tool timeline, token metrics.
 * Port of apps/syn-cli/src/syn_cli/commands/observe.py
 */

import { CommandGroup, type CommandDef, type ParsedArgs } from "../framework/command.js";
import { CLIError } from "../framework/errors.js";
import { api, unwrap } from "../client/typed.js";
import type { components } from "../generated/api-types.js";
import { print, printError, printDim } from "../output/console.js";
import { style, BOLD, CYAN, DIM, GREEN, RED } from "../output/ansi.js";
import { formatCost, formatDuration, formatTimestamp } from "../output/format.js";
import { Table } from "../output/table.js";

type ToolTimeline = components["schemas"]["ToolTimelineResponse"];
type ToolTimelineEntry = components["schemas"]["ToolTimelineEntry"];
type TokenMetrics = components["schemas"]["SessionTokenMetrics"];

function reqSessionId(parsed: ParsedArgs): string {
  const id = parsed.positionals[0];
  if (!id) { printError("Missing session-id"); throw new CLIError("Missing argument", 1); }
  return id;
}

const toolTimelineCommand: CommandDef = {
  name: "tools",
  description: "Show tool execution timeline for a session",
  args: [{ name: "session-id", description: "Session ID", required: true }],
  options: {
    limit: { type: "string", description: "Max entries", default: "100" },
  },
  handler: async (parsed: ParsedArgs) => {
    const sid = reqSessionId(parsed);
    const limitStr = (parsed.values["limit"] as string | undefined) ?? "100";

    const data: ToolTimeline = unwrap(await api.GET("/observability/sessions/{session_id}/tools", {
      params: {
        path: { session_id: sid },
        query: { limit: parseInt(limitStr, 10) },
      },
    }), "Failed to fetch tool timeline");

    const entries: ToolTimelineEntry[] = data.executions ?? [];
    if (entries.length === 0) { printDim("No tool timeline entries."); return; }

    const table = new Table({ title: `Tool Timeline: ${sid.slice(0, 12)}` });
    table.addColumn("Time");
    table.addColumn("Tool", { style: CYAN });
    table.addColumn("Duration", { align: "right" });
    table.addColumn("Status");

    for (const e of entries) {
      table.addRow(
        formatTimestamp(String(e.timestamp ?? "")),
        e.tool_name ?? "",
        e.duration_ms != null ? formatDuration(e.duration_ms) : "\u2014",
        e.success === true ? style("ok", GREEN) : e.success === false ? style("error", RED) : style("\u2014", DIM),
      );
    }
    table.print();
  },
};

const tokenMetricsCommand: CommandDef = {
  name: "tokens",
  description: "Show token breakdown for a session",
  args: [{ name: "session-id", description: "Session ID", required: true }],
  handler: async (parsed: ParsedArgs) => {
    const sid = reqSessionId(parsed);

    const d: TokenMetrics = unwrap(await api.GET("/observability/sessions/{session_id}/tokens", {
      params: { path: { session_id: sid } },
    }), "Failed to fetch token metrics");

    print(`${style("Token Metrics:", BOLD)} ${d.session_id}`);
    print(`  Input tokens:       ${d.input_tokens.toLocaleString()}`);
    print(`  Output tokens:      ${d.output_tokens.toLocaleString()}`);
    print(`  Total tokens:       ${d.total_tokens.toLocaleString()}`);
    if (d.cache_creation_tokens) print(`  Cache creation:     ${d.cache_creation_tokens.toLocaleString()}`);
    if (d.cache_read_tokens) print(`  Cache read:         ${d.cache_read_tokens.toLocaleString()}`);
    if (d.total_cost_usd !== "0")
      print(`  Estimated cost:     ${formatCost(d.total_cost_usd)}`);
  },
};

export const observeGroup = new CommandGroup("observe", "Session observability: tool timelines and token metrics");
observeGroup.command(toolTimelineCommand).command(tokenMetricsCommand);
